Match returns null with no matches

runn1ng opened the pull request:
"foobar".match(/Foo/) returns null.

See [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match)

> Returns
>
> array
> 
> An Array containing the matched results or null if there were no matches.

runn1ng commented:
...well, all the tests failed, because `match()` result can now be null. I don't feel qualified enough to fix them.

samwgoldman commented:
@runn1ng Those errors aren't too bad! I think you can do it.

If you aren't familiar with the test set up, it's not too hard to explain.

If there is a directory under `test`, like `test/example`, then there will be a file called `test/example/example.exp` that includes the expected output from `flow check run against the `.js` files in `test/example`.

The test runner will run `flow check` and redirect the output to `test/example/example.out` and generate a diff between the `.exp` and the `.out` file into `test/example/example.diff`. If the diff is non-empty, the test "failed."

It is possible (and quite easy) to make a change that creates a diff, but does not represent an actual test failure. In this case, you can use patch to update the `.exp` file, like so: `patch test/example/example.ext test/example/example.diff`.

OK, for these failures, here's what I would do:

1. For each error, I would use the filename/lineno to locate the code in the test file.
2. If it looks like the return type of the match is part of the assertion, I would change the test to first do a dynamic type test for null, then proceed with the normal check. (Otherwise, just patch and you're done.)
3. This will probably change some line numbers, so run again, and if only line numbers are different, re-patch.

Unfortunately, some of the tests are not as clear as they could be about what they mean/expect, so if you are uncertain about that, feel free to paste it here and I (or someone else) will help you get unstuck!

runn1ng commented:
Haha, all right, let's try this.

gabelevi commented:
Looks like the last failing test is just because it assumes that `match()` returns an array. If you update the [test](https://github.com/facebook/flow/blob/de990dea17873393d1d91ad10255d6dc0c638a27/tests/overload/overload.js#L1) you should be set

samwgoldman commented:
@gabelevi Do you think we can take this? I know you can make the necessary changes during your import workflow. (I'd do it myself, but I'm on the outside looking in here.)

runn1ng commented:
Sorry that I left this PR rot, I don't really have time to look at it :/

runn1ng commented:
Well I fixed it by modifying the expected values, not the test itself. (Plus I rebased the branch on master, but that was without any problems)

Is that OK?

gabelevi commented:
Sweet, thanks for updating this PR! Looks good to me now!

gabelevi commented:
@facebook-github-bot shipit

gabelevi commented:
Hmm, that didn't seem to work...let's try one more time...

gabelevi commented:
@facebook-github-bot shipit
